content/tutorial-deep-learning-on-mnist.ipynb

```diff
@@ -0,0 +1,1017 @@
+{
```

Comment:
<ul><li>Why do you test every 10 epochs? </li><li>Is it necessary to use <code>[i, i+1]</code>  for each item, or could <code>[i]</code>  do the same job?</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/numpy/numpy-tutorials/pull/33/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='67'/>

Reply:
- I think in the original gist, you test (evaluate) the model every 10 epochs because it should have more "knowledge" to be tested after the 10th training iteration. I've tried testing at every epoch and noticed the results were OK in this simple example. It's reflected in the new code.
- `[i:i+1]` is necessary here (doesn't seem to work without it)

Reply:
<ul><li>Why do you print "Epoch" + str(j+10)? I think there is no need to add 10 to this result - am I missing something?</li><li>Ok I got the reason for the indexing but this code could be simplified as follows:</li></ul><pre class="ql-syntax" spellcheck="false"># To store training and test set losses and accurate predictions
# for visualization.
store_training_loss = []
store_training_accurate_pred = []
store_test_loss = []
store_test_accurate_pred = []

# This is a training loop.
# Run the learning experiment for a defined number of epochs (iterations).
for j in range(epochs):
&nbsp;&nbsp;&nbsp; # Set the initial loss/error and the number of accurate predictions to zero.
&nbsp;&nbsp;&nbsp; training_loss = 0.0
&nbsp;&nbsp;&nbsp; training_accurate_predictions = 0
&nbsp;&nbsp; &nbsp;
&nbsp;&nbsp;&nbsp; # For all images in the training set, perform a forward pass
&nbsp;&nbsp;&nbsp; # and backpropagation and adjust the weights accordingly.
&nbsp;&nbsp;&nbsp; for i in range(len(training_images)):
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # Forward propagation/forward pass:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 1. The input layer:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; #&nbsp;&nbsp;&nbsp; Initialize the training image data as inputs.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; layer_0 = training_images[i]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 2. The hidden layer:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; #&nbsp;&nbsp;&nbsp; Take in the training image data into the middle layer by
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; #&nbsp;&nbsp;&nbsp; matrix-multiplying it by randomly initialized weights.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; layer_1 = np.dot(layer_0, weights_1)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 3. Pass the hidden layer's output through the ReLU activation function.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; layer_1 = relu(layer_1)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 4. Define the dropout function for regularization.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; dropout_mask = np.random.randint(0, high=2, size=layer_1.shape)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 5. Apply dropout to the hidden layer's output.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; layer_1 *= dropout_mask * 2
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 6. The output layer:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; #&nbsp;&nbsp;&nbsp; Ingest the output of the middle layer into the the final layer
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; #&nbsp;&nbsp;&nbsp; by matrix-multiplying it by randomly initialized weights.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; #&nbsp;&nbsp;&nbsp; Produce a 10-dimension vector with 10 scores.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; layer_2 = np.dot(layer_1, weights_2)

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # Backpropagation/backward pass:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 1. Measure the training error (loss function) between the actual
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; #&nbsp;&nbsp;&nbsp; image labels (the truth) and the prediction by the model.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; training_loss += np.sum((training_labels[i] - layer_2) ** 2)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 2. Increment the accurate prediction count.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; training_accurate_predictions += int(np.argmax(layer_2) == np.argmax(training_labels[i]))
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 3. Differentiate the loss function/error.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; layer_2_delta = (training_labels[i] - layer_2)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 4. Propagate the gradients of the loss function back through the hidden layer.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; layer_1_delta = np.dot(weights_2, layer_2_delta) * relu2deriv(layer_1)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 5. Apply the dropout to the gradients.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; layer_1_delta *= dropout_mask
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 6. Update the weights for the middle and input layers
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; #&nbsp;&nbsp;&nbsp; by multiplying them by the learning rate and the gradients.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; weights_2 += learning_rate * np.outer(layer_1, layer_2_delta)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; weights_1 += learning_rate * np.outer(layer_0, layer_1_delta)
&nbsp;&nbsp; &nbsp;
&nbsp;&nbsp;&nbsp; # Store training set losses and accurate predictions.
&nbsp;&nbsp;&nbsp; store_training_loss.append(training_loss)
&nbsp;&nbsp;&nbsp; store_training_accurate_pred.append(training_accurate_predictions)

&nbsp;&nbsp;&nbsp; # Evaluate on the test set:
&nbsp;&nbsp;&nbsp; # 1. Set the initial error and the number of accurate predictions to zero.
&nbsp;&nbsp;&nbsp; test_loss = 0.0
&nbsp;&nbsp;&nbsp; test_accurate_predictions = 0
&nbsp;&nbsp; &nbsp;
&nbsp;&nbsp;&nbsp; # 2. Start testing the model by evaluating on the test image dataset.
&nbsp;&nbsp;&nbsp; for i in range(len(test_images)):
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 1. Pass the test images through the input layer.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; layer_0 = test_images[i]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 2. Compute the weighted sum of the test image inputs in and
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; #&nbsp;&nbsp;&nbsp; pass the hidden layer's output through ReLU.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; layer_1 = relu(np.dot(layer_0, weights_1))
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 3. Compute the weighted sum of the hidden layer's inputs.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; #&nbsp;&nbsp;&nbsp; Produce a 10-dimensional vector with 10 scores.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; layer_2 = np.dot(layer_1, weights_2)

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 4. Measure the error between the actual label (truth) and prediction values.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; test_loss += np.sum((test_labels[i] - layer_2) ** 2)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # 5. Increment the accurate prediction count.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; test_accurate_predictions += int(np.argmax(layer_2) == np.argmax(test_labels[i]))

&nbsp;&nbsp;&nbsp; # Store test set losses and accurate predictions.
&nbsp;&nbsp;&nbsp; store_test_loss.append(test_loss)
&nbsp;&nbsp;&nbsp; store_test_accurate_pred.append(test_accurate_predictions)

&nbsp;&nbsp;&nbsp; # 3. Display the error and accuracy metrics in the output.
&nbsp;&nbsp;&nbsp; print("\n" + \
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; "Epoch: " + str(j) + \
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; " Training set error:" + str(training_loss/ float(len(training_images)))[0:5] +\
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; " Training set accuracy:" + str(training_accurate_predictions/ float(len(training_images))) +\
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; " Test set error:" + str(test_loss/ float(len(test_images)))[0:5] +\
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; " Test set accuracy:" + str(test_accurate_predictions/ float(len(test_images))))
</pre><p>Let me know if this makes sense to you. There was nothing wrong with the original code - I just think it is easier to avoid transposing if you can, and using 1d arrays instead of 2d. </p>

Reply:
The epoch number used to be printed every 10th time - it can be less overwhelming as you increase the epochs to thousands. 

I suggest we use your simplified version of the code and write `"Epoch: " + str(j+1)` since we start at `0`, so the count ends at `100` instead of `99`.

I checked the "before" and "after" error and accuracy rates and they appear to be the same — not great (because this is a simple example) but relatively consistent.

Thanks for the revised code! 🙏